lms/models/groups.py

```diff
@@ -0,0 +1,53 @@
+# -*- coding: utf-8 -*-
+
+import sqlalchemy as sa
+
+from lms.db import BASE
+
+
+class Group(BASE):
+    """A record of a group that was successfully created in h."""
+
+    __tablename__ = "groups"
+    __table_args__ = (
+        # Add a composite index of the (tool_consumer_instance_guid, context_id)
+        # columns. This should make looking up groups table rows by
+        # (tool_consumer_instance_guid, context_id), which we're expecting to do
+        # often, faster.
+        #
+        # See:
+        #
+        # * http://docs.sqlalchemy.org/en/latest/core/constraints.html#indexes
+        # * https://www.postgresql.org/docs/10/static/indexes.html
+        sa.Index(
+            "ix__groups_tool_consumer_instance_guid_context_id",
+            "tool_consumer_instance_guid",
+            "context_id",
+            # Add a uniqueness constraint to the index so that two groups can't
+            # have the same tool_consumer_instance_guid and context_id.
+            unique=True,
+        ),
+    )
+
+    id = sa.Column(sa.Integer, autoincrement=True, primary_key=True)
+    tool_consumer_instance_guid = sa.Column(sa.UnicodeText, nullable=False)
+    context_id = sa.Column(sa.UnicodeText, nullable=False)
+    pubid = sa.Column(sa.Text, nullable=False, unique=True)
+
+    @classmethod
+    def get(cls, db, tool_consumer_instance_guid, context_id):
+        """
+        Return the Group with the given properties, or None.
+
+        Return the Group with the given tool_consumer_instance_guid and
+        context_id, or None.
+
+        """
+        return (
+            db.query(cls)
+            .filter_by(
+                tool_consumer_instance_guid=tool_consumer_instance_guid,
+                context_id=context_id,
+            )
+            .one_or_none()
+        )
```

Comment:
Here we're encapsulating sqalchemy in the models. Code that wants to use the groups table in the DB doesn't need to use sqlalchemy at all (well, other than calling `request.db.add()` to add groups to the DB, and `request.db.delete()` to delete them). Instead it just works with our `Group` class's simple (and easily mockable) public API: constructing `Group` objects ) and the `tool_consumer_instance_guid`, `context_id`, and `pubid` arguments to `__init__()`), reading and writing the `.tool_consumer_instance_guid`, `.context_id`, and `.pubid` properties of group objects, and calling the simple `Group.get()` method to get groups from the DB. No sqlalchemy queries outside of the models.